File: src/main/java/fromjavatokotlin/model/Robot.java
package fromjavatokotlin.model;

import lombok.*;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.util.ArrayList;
import java.util.List;

@Document("robots")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@ToString(callSuper = true)
public class Robot extends Identifiable {

  public static final String AVAILABILITY_FIELD = "availability";
  public static final String ABILITIES_FIELD = "abilities";

  private String code;

  private Integer mass;

  @Field(AVAILABILITY_FIELD)
  private Boolean availability;

  @Field(ABILITIES_FIELD)
  private List<Ability> abilities = new ArrayList<>();

  @Getter
  @Setter
  @AllArgsConstructor
  @NoArgsConstructor
  @ToString
  @EqualsAndHashCode
  public static class Ability {
    private String code;
  }

}

File: src/main/kotlin/fromjavatokotlin/model/Robot.kt
package fromjavatokotlin.model

import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.Field
import java.util.*

@Document("robots")
class Robot(
    val code: String,
    val mass: Int? = null,
    @Field(AVAILABILITY_FIELD)
    val availability: Boolean? = null,
    @Field(ABILITIES_FIELD)
    val abilities: List<Ability> = ArrayList()
) : Identifiable() {

    data class Ability(var code: String? = null)

    companion object {
        const val AVAILABILITY_FIELD = "availability"
        const val ABILITIES_FIELD = "abilities"
    }
}